ICE: assertion failed: !suggestion.is_empty()

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
struct Bug<S> {
    A: [(); {
        let x: [u8; Self::W] = [0; Self::W];
        F
    }
}

fn main() {}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.61.0-nightly (10913c000 2022-03-03)
binary: rustc
commit-hash: 10913c00018c76103b2fd4260d8c02ec728fd244
commit-date: 2022-03-03
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Error output

```
error: mismatched closing delimiter: `}`
 --> icemaker_175.rs:3:8
  |
2 | struct Bug<S> {
  |               - closing delimiter possibly meant for this
3 |     A: [(); {
  |        ^ unclosed delimiter
...
7 | }
  | ^ mismatched closing delimiter

error: generic parameters may not be used in const operations
  --> icemaker_175.rs:11:19
   |
11 |     fn d(r: &[u8; Self::W]) -> Self;
   |                   ^^^^^^^ cannot perform const operation using `Self`
   |
   = note: type parameters may not be used in const expressions
   = help: use `#![feature(generic_const_exprs)]` to allow generic const expressions

error[E0425]: cannot find value `F` in this scope
 --> icemaker_175.rs:5:9
  |
5 |         F
  |         ^
  |
help: a local variable with a similar name exists
  |
5 |         x
  |
help: you might be missing a type parameter
  |
2 | struct Bug<S, F> {
  |             +++

error: generic `Self` types are currently not permitted in anonymous constants
 --> icemaker_175.rs:4:21
  |
4 |         let x: [u8; Self::W] = [0; Self::W];
  |                     ^^^^

error: generic `Self` types are currently not permitted in anonymous constants
 --> icemaker_175.rs:4:36
  |
4 |         let x: [u8; Self::W] = [0; Self::W];
  |                                    ^^^^

thread 'rustc' panicked at 'assertion failed: !suggestion.is_empty()', compiler/rustc_errors/src/diagnostic.rs:399:9
stack backtrace:
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'assertion failed: !suggestion.is_empty()', compiler/rustc_errors/src/diagnostic.rs:399:9
stack backtrace:
   0:     0x7f829509cc5c - std::backtrace_rs::backtrace::libunwind::trace::hf0766e62d7eade0a
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f829509cc5c - std::backtrace_rs::backtrace::trace_unsynchronized::hfeabd612f0ca832b
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f829509cc5c - std::sys_common::backtrace::_print_fmt::he9db73d5ee638646
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f829509cc5c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1589ebb97e621cc4
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f82950f7dac - core::fmt::write::ha37bcdb331738e85
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/core/src/fmt/mod.rs:1190:17
   5:     0x7f829508d781 - std::io::Write::write_fmt::h718204eaedb628ac
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/io/mod.rs:1657:15
   6:     0x7f829509fbc5 - std::sys_common::backtrace::_print::ha178300e1fe9607f
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f829509fbc5 - std::sys_common::backtrace::print::hc7007f5e7ae2c2cd
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f829509fbc5 - std::panicking::default_hook::{{closure}}::he22cced35b2db4f8
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/panicking.rs:295:22
   9:     0x7f829509f874 - std::panicking::default_hook::hc232514524fdf7f9
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/panicking.rs:314:9
  10:     0x7f82958c5561 - rustc_driver[df60e0066a970eb8]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f82950a044f - std::panicking::rust_panic_with_hook::h8f622f46b9144300
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/panicking.rs:702:17
  12:     0x7f82950a00f9 - std::panicking::begin_panic_handler::{{closure}}::h12413d9df30e1a50
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/panicking.rs:586:13
  13:     0x7f829509d114 - std::sys_common::backtrace::__rust_end_short_backtrace::hb3ebc8bc50cc5679
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7f829509fe69 - rust_begin_unwind
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/panicking.rs:584:5
  15:     0x7f8295064b63 - core::panicking::panic_fmt::hbbe37082ef5d13e7
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/core/src/panicking.rs:143:14
  16:     0x7f8295064a2d - core::panicking::panic::hfd6ebb794049fd71
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/core/src/panicking.rs:48:5
  17:     0x7f8297fe19ff - <rustc_errors[1a4bb3cbefced23b]::diagnostic::Diagnostic>::multipart_suggestion_with_style
  18:     0x7f8297fe183b - <rustc_errors[1a4bb3cbefced23b]::diagnostic::Diagnostic>::multipart_suggestion_verbose
  19:     0x7f82968d61ee - rustc_middle[d2c2c4d42ddfe9db]::ty::diagnostics::suggest_constraining_type_param
  20:     0x7f829669ebae - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxt as rustc_trait_selection[33828acc4ff396a4]::traits::error_reporting::suggestions::InferCtxtExt>::suggest_restricting_param_bound
  21:     0x7f82966b473b - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxt as rustc_trait_selection[33828acc4ff396a4]::traits::error_reporting::InferCtxtExt>::report_selection_error
  22:     0x7f82966bb08b - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxt as rustc_trait_selection[33828acc4ff396a4]::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  23:     0x7f82966b0825 - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxt as rustc_trait_selection[33828acc4ff396a4]::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  24:     0x7f8296c5ea45 - <rustc_typeck[90eb3e766239d175]::check::fn_ctxt::FnCtxt>::type_inference_fallback
  25:     0x7f8296cf6e85 - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxtBuilder>::enter::<&rustc_middle[d2c2c4d42ddfe9db]::ty::context::TypeckResults, <rustc_typeck[90eb3e766239d175]::check::inherited::InheritedBuilder>::enter<rustc_typeck[90eb3e766239d175]::check::typeck_with_fallback<rustc_typeck[90eb3e766239d175]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[d2c2c4d42ddfe9db]::ty::context::TypeckResults>::{closure#0}>
  26:     0x7f8296cdfcce - rustc_typeck[90eb3e766239d175]::check::typeck
  27:     0x7f829718eb8f - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::LocalDefId, &rustc_middle[d2c2c4d42ddfe9db]::ty::context::TypeckResults>>
  28:     0x7f82972706e8 - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::typeck
  29:     0x7f82975ed25f - <rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt>::typeck_opt_const_arg
  30:     0x7f8296ed0d4b - rustc_mir_build[3886015364474496]::build::mir_built
  31:     0x7f829717d334 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::WithOptConstParam<rustc_span[c03518380bcf3f86]::def_id::LocalDefId>, &rustc_data_structures[a60ea62ea8c2b9e5]::steal::Steal<rustc_middle[d2c2c4d42ddfe9db]::mir::Body>>>
  32:     0x7f829726ee5b - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_built
  33:     0x7f8296c0787d - rustc_mir_transform[f2cb810e1b99ca21]::check_unsafety::unsafety_check_result
  34:     0x7f8296c03a88 - <rustc_mir_transform[f2cb810e1b99ca21]::check_unsafety::provide::{closure#0} as core[a9132578ce6008ac]::ops::function::FnOnce<(rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt, rustc_span[c03518380bcf3f86]::def_id::LocalDefId)>>::call_once
  35:     0x7f8297190bdf - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::LocalDefId, &rustc_middle[d2c2c4d42ddfe9db]::mir::query::UnsafetyCheckResult>>
  36:     0x7f82972702ee - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::unsafety_check_result
  37:     0x7f8296c1020d - rustc_mir_transform[f2cb810e1b99ca21]::mir_const
  38:     0x7f829717d334 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::WithOptConstParam<rustc_span[c03518380bcf3f86]::def_id::LocalDefId>, &rustc_data_structures[a60ea62ea8c2b9e5]::steal::Steal<rustc_middle[d2c2c4d42ddfe9db]::mir::Body>>>
  39:     0x7f829726ef92 - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_const
  40:     0x7f8296c02beb - <rustc_mir_transform[f2cb810e1b99ca21]::provide::{closure#0} as core[a9132578ce6008ac]::ops::function::FnOnce<(rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt, rustc_span[c03518380bcf3f86]::def_id::DefId)>>::call_once
  41:     0x7f829719e982 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::DefId, rustc_middle[d2c2c4d42ddfe9db]::mir::query::ConstQualifs>>
  42:     0x7f829726ecfe - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_const_qualif
  43:     0x7f8296c110c5 - rustc_mir_transform[f2cb810e1b99ca21]::mir_promoted
  44:     0x7f829717e902 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::WithOptConstParam<rustc_span[c03518380bcf3f86]::def_id::LocalDefId>, (&rustc_data_structures[a60ea62ea8c2b9e5]::steal::Steal<rustc_middle[d2c2c4d42ddfe9db]::mir::Body>, &rustc_data_structures[a60ea62ea8c2b9e5]::steal::Steal<rustc_index[ba9440815735119c]::vec::IndexVec<rustc_middle[d2c2c4d42ddfe9db]::mir::Promoted, rustc_middle[d2c2c4d42ddfe9db]::mir::Body>>)>>
  45:     0x7f829726f39b - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_promoted
  46:     0x7f8296feea51 - rustc_borrowck[c32371fceb3e7611]::mir_borrowck
  47:     0x7f8296feb8d8 - <rustc_borrowck[c32371fceb3e7611]::provide::{closure#0} as core[a9132578ce6008ac]::ops::function::FnOnce<(rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt, rustc_span[c03518380bcf3f86]::def_id::LocalDefId)>>::call_once
  48:     0x7f829718fbbc - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::LocalDefId, &rustc_middle[d2c2c4d42ddfe9db]::mir::query::BorrowCheckResult>>
  49:     0x7f829727095a - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_borrowck
  50:     0x7f82975ed7f5 - <rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt>::mir_borrowck_opt_const_arg
  51:     0x7f8296c1342d - rustc_mir_transform[f2cb810e1b99ca21]::mir_drops_elaborated_and_const_checked
  52:     0x7f829717d334 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::WithOptConstParam<rustc_span[c03518380bcf3f86]::def_id::LocalDefId>, &rustc_data_structures[a60ea62ea8c2b9e5]::steal::Steal<rustc_middle[d2c2c4d42ddfe9db]::mir::Body>>>
  53:     0x7f829726f25e - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::mir_drops_elaborated_and_const_checked
  54:     0x7f829791ec1f - rustc_mir_transform[f2cb810e1b99ca21]::inner_mir_for_ctfe
  55:     0x7f829791e483 - rustc_mir_transform[f2cb810e1b99ca21]::mir_for_ctfe
  56:     0x7f82971a0cb2 - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::DefId, &rustc_middle[d2c2c4d42ddfe9db]::mir::Body>>
  57:     0x7f8297c3606c - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::get_query::<rustc_query_impl[9f92f053cfcb6197]::queries::mir_for_ctfe, rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt>
  58:     0x7f82960c485d - <rustc_const_eval[3800a605cb14db8d]::const_eval::machine::CompileTimeInterpreter as rustc_const_eval[3800a605cb14db8d]::interpret::machine::Machine>::load_mir
  59:     0x7f8297b74716 - <rustc_const_eval[3800a605cb14db8d]::interpret::eval_context::InterpCx<rustc_const_eval[3800a605cb14db8d]::const_eval::machine::CompileTimeInterpreter>>::load_mir
  60:     0x7f829714f222 - rustc_const_eval[3800a605cb14db8d]::const_eval::eval_queries::eval_to_allocation_raw_provider
  61:     0x7f8297178b2d - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::ParamEnvAnd<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::GlobalId>, core[a9132578ce6008ac]::result::Result<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::value::ConstAlloc, rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::error::ErrorHandled>>>
  62:     0x7f8297270aa3 - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::eval_to_allocation_raw
  63:     0x7f829714e0b8 - rustc_const_eval[3800a605cb14db8d]::const_eval::eval_queries::eval_to_const_value_raw_provider
  64:     0x7f8297179c2a - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::ParamEnvAnd<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::GlobalId>, core[a9132578ce6008ac]::result::Result<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::value::ConstValue, rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::error::ErrorHandled>>>
  65:     0x7f8297270c4f - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::eval_to_const_value_raw
  66:     0x7f829714ddf4 - rustc_const_eval[3800a605cb14db8d]::const_eval::eval_queries::eval_to_const_value_raw_provider
  67:     0x7f8297179c2a - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_middle[d2c2c4d42ddfe9db]::ty::ParamEnvAnd<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::GlobalId>, core[a9132578ce6008ac]::result::Result<rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::value::ConstValue, rustc_middle[d2c2c4d42ddfe9db]::mir::interpret::error::ErrorHandled>>>
  68:     0x7f8297270c4f - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::eval_to_const_value_raw
  69:     0x7f82975e1ecd - <rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt>::const_eval_global_id
  70:     0x7f8297f59389 - <rustc_middle[d2c2c4d42ddfe9db]::ty::context::TyCtxt>::const_eval_resolve
  71:     0x7f8297ecc836 - <rustc_trait_selection[33828acc4ff396a4]::traits::project::AssocTypeNormalizer as rustc_middle[d2c2c4d42ddfe9db]::ty::fold::TypeFolder>::fold_const
  72:     0x7f82975163f8 - <rustc_middle[d2c2c4d42ddfe9db]::ty::Ty as rustc_middle[d2c2c4d42ddfe9db]::ty::fold::TypeFoldable>::super_fold_with::<rustc_trait_selection[33828acc4ff396a4]::traits::project::AssocTypeNormalizer>
  73:     0x7f82974ac458 - <rustc_trait_selection[33828acc4ff396a4]::traits::project::AssocTypeNormalizer as rustc_middle[d2c2c4d42ddfe9db]::ty::fold::TypeFolder>::fold_ty
  74:     0x7f8296c60d86 - <rustc_typeck[90eb3e766239d175]::check::fn_ctxt::FnCtxt>::normalize_associated_types_in::<rustc_middle[d2c2c4d42ddfe9db]::ty::Ty>
  75:     0x7f8297a33b6d - <core[a9132578ce6008ac]::iter::adapters::map::Map<core[a9132578ce6008ac]::slice::iter::Iter<rustc_hir[ac8a1b90366884e9]::hir::FieldDef>, <rustc_typeck[90eb3e766239d175]::check::fn_ctxt::FnCtxt>::non_enum_variant::{closure#0}> as core[a9132578ce6008ac]::iter::traits::iterator::Iterator>::fold::<(), core[a9132578ce6008ac]::iter::traits::iterator::Iterator::for_each::call<rustc_typeck[90eb3e766239d175]::check::wfcheck::AdtField, <alloc[ca008d8e5ee5d252]::vec::Vec<rustc_typeck[90eb3e766239d175]::check::wfcheck::AdtField> as alloc[ca008d8e5ee5d252]::vec::spec_extend::SpecExtend<rustc_typeck[90eb3e766239d175]::check::wfcheck::AdtField, core[a9132578ce6008ac]::iter::adapters::map::Map<core[a9132578ce6008ac]::slice::iter::Iter<rustc_hir[ac8a1b90366884e9]::hir::FieldDef>, <rustc_typeck[90eb3e766239d175]::check::fn_ctxt::FnCtxt>::non_enum_variant::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
  76:     0x7f8297a28119 - <alloc[ca008d8e5ee5d252]::vec::Vec<rustc_typeck[90eb3e766239d175]::check::wfcheck::AdtField> as alloc[ca008d8e5ee5d252]::vec::spec_from_iter::SpecFromIter<rustc_typeck[90eb3e766239d175]::check::wfcheck::AdtField, core[a9132578ce6008ac]::iter::adapters::map::Map<core[a9132578ce6008ac]::slice::iter::Iter<rustc_hir[ac8a1b90366884e9]::hir::FieldDef>, <rustc_typeck[90eb3e766239d175]::check::fn_ctxt::FnCtxt>::non_enum_variant::{closure#0}>>>::from_iter
  77:     0x7f82979de46c - <rustc_infer[36cd1921c7e1305c]::infer::InferCtxtBuilder>::enter::<(), <rustc_typeck[90eb3e766239d175]::check::inherited::InheritedBuilder>::enter<<rustc_typeck[90eb3e766239d175]::check::wfcheck::CheckWfFcxBuilder>::with_fcx<rustc_typeck[90eb3e766239d175]::check::wfcheck::check_type_defn<rustc_typeck[90eb3e766239d175]::check::wfcheck::check_item_well_formed::{closure#2}>::{closure#0}>::{closure#0}, ()>::{closure#0}>
  78:     0x7f8296d368bd - rustc_typeck[90eb3e766239d175]::check::wfcheck::check_item_well_formed
  79:     0x7f8297191c0f - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<rustc_span[c03518380bcf3f86]::def_id::LocalDefId, ()>>
  80:     0x7f8297272541 - <rustc_query_impl[9f92f053cfcb6197]::Queries as rustc_middle[d2c2c4d42ddfe9db]::ty::query::QueryEngine>::check_item_well_formed
  81:     0x7f8296d3fc21 - <rustc_middle[d2c2c4d42ddfe9db]::hir::map::Map>::par_visit_all_item_likes::<rustc_typeck[90eb3e766239d175]::check::wfcheck::CheckTypeWellFormedVisitor>
  82:     0x7f82979d1567 - <rustc_session[1ec8f008caf3efe6]::session::Session>::track_errors::<rustc_typeck[90eb3e766239d175]::check_crate::{closure#5}, ()>
  83:     0x7f82979ce6ab - rustc_typeck[90eb3e766239d175]::check_crate
  84:     0x7f82977553d7 - rustc_interface[3ed5997e01d12d63]::passes::analysis
  85:     0x7f8297c0b77b - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::try_execute_query::<rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt, rustc_query_system[12c6c57c6519fd3c]::query::caches::DefaultCache<(), core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>>
  86:     0x7f8297c4a5fe - rustc_query_system[12c6c57c6519fd3c]::query::plumbing::get_query::<rustc_query_impl[9f92f053cfcb6197]::queries::analysis, rustc_query_impl[9f92f053cfcb6197]::plumbing::QueryCtxt>
  87:     0x7f829773cca4 - <rustc_interface[3ed5997e01d12d63]::passes::QueryContext>::enter::<rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>
  88:     0x7f829771a57a - <rustc_interface[3ed5997e01d12d63]::interface::Compiler>::enter::<rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}::{closure#2}, core[a9132578ce6008ac]::result::Result<core[a9132578ce6008ac]::option::Option<rustc_interface[3ed5997e01d12d63]::queries::Linker>, rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>
  89:     0x7f829772d3ad - rustc_span[c03518380bcf3f86]::with_source_map::<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_interface[3ed5997e01d12d63]::interface::create_compiler_and_run<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}>::{closure#1}>
  90:     0x7f829771b490 - rustc_interface[3ed5997e01d12d63]::interface::create_compiler_and_run::<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}>
  91:     0x7f8297719432 - <scoped_tls[651cb9229c2f64c4]::ScopedKey<rustc_span[c03518380bcf3f86]::SessionGlobals>>::set::<rustc_interface[3ed5997e01d12d63]::interface::run_compiler<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}>::{closure#0}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>
  92:     0x7f829771771d - std[d2b06600a978a9da]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3ed5997e01d12d63]::util::run_in_thread_pool_with_globals<rustc_interface[3ed5997e01d12d63]::interface::run_compiler<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}>::{closure#0}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>::{closure#0}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>
  93:     0x7f8297717a09 - <<std[d2b06600a978a9da]::thread::Builder>::spawn_unchecked_<rustc_interface[3ed5997e01d12d63]::util::run_in_thread_pool_with_globals<rustc_interface[3ed5997e01d12d63]::interface::run_compiler<core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>, rustc_driver[df60e0066a970eb8]::run_compiler::{closure#1}>::{closure#0}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>::{closure#0}, core[a9132578ce6008ac]::result::Result<(), rustc_errors[1a4bb3cbefced23b]::ErrorGuaranteed>>::{closure#1} as core[a9132578ce6008ac]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  94:     0x7f82950aa9d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he4379d5c274940c6
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/alloc/src/boxed.rs:1854:9
  95:     0x7f82950aa9d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h2db866a59e058c86
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/alloc/src/boxed.rs:1854:9
  96:     0x7f82950aa9d3 - std::sys::unix::thread::Thread::new::thread_start::h2eda9002b1b74dd5
                               at /rustc/10913c00018c76103b2fd4260d8c02ec728fd244/library/std/src/sys/unix/thread.rs:108:17
  97:     0x7f8294e835c2 - start_thread
  98:     0x7f8294f08584 - __clone
  99:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.61.0-nightly (10913c000 2022-03-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `Bug::A::{constant#0}`
#1 [mir_built] building MIR for `Bug::A::{constant#0}`
#2 [unsafety_check_result] unsafety-checking `Bug::A::{constant#0}`
#3 [mir_const] processing MIR for `Bug::A::{constant#0}`
#4 [mir_const_qualif] const checking `Bug::A::{constant#0}`
#5 [mir_promoted] processing `Bug::A::{constant#0}`
#6 [mir_borrowck] borrow-checking `Bug::A::{constant#0}`
#7 [mir_drops_elaborated_and_const_checked] elaborating drops for `Bug::A::{constant#0}`
#8 [mir_for_ctfe] caching mir of `Bug::A::{constant#0}` for CTFE
#9 [eval_to_allocation_raw] const-evaluating + checking `Bug::A::{constant#0}`
#10 [eval_to_const_value_raw] simplifying constant for the type system `Bug::A::{constant#0}`
#11 [eval_to_const_value_raw] simplifying constant for the type system `Bug::A::{constant#0}`
#12 [check_item_well_formed] checking that `Bug` is well-formed
#13 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0425`.
```

</p>
</details>



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"ouz-a"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_middle/src/ty/diagnostics.rs
+++ b/compiler/rustc_middle/src/ty/diagnostics.rs
@@ -325,12 +325,6 @@ pub fn suggest_constraining_type_params<'a>(
             }
         }
 
-        // This check is always run on non-valid code
-        // to not trigger ICE
-        if constraints.is_empty() && suggestions.is_empty() {
-            return false;
-        }
-
         if constraints.is_empty() {
             continue;
         }
@@ -518,7 +512,7 @@ pub fn suggest_constraining_type_params<'a>(
         };
 
         err.span_suggestion_verbose(span, msg, suggestion, applicability);
-    } else {
+    } else if suggestions.len() > 1 {
         err.multipart_suggestion_verbose(
             "consider restricting type parameters",
             suggestions.into_iter().map(|(span, suggestion, _)| (span, suggestion)).collect(),

--- a/compiler/rustc_middle/src/ty/diagnostics.rs
+++ b/compiler/rustc_middle/src/ty/diagnostics.rs
@@ -325,6 +325,12 @@ pub fn suggest_constraining_type_params<'a>(
             }
         }
 
+        // This check is always run on non-valid code
+        // to not trigger ICE
+        if constraints.is_empty() && suggestions.is_empty() {
+            return false;
+        }
+
         if constraints.is_empty() {
             continue;
         }

--- a/src/test/ui/suggestions/constrain-suggest-ice.rs
+++ b/src/test/ui/suggestions/constrain-suggest-ice.rs
@@ -1,0 +1,11 @@
+struct Bug<S>{ //~ ERROR parameter `S` is never used [E0392]
+    A: [(); {
+        let x: [u8; Self::W] = [0; Self::W]; //~ ERROR generic `Self` types are currently not permitted in anonymous constants
+        //~^ ERROR generic `Self` types are currently not permitted in anonymous constants
+        //~^^ ERROR the size for values of type `S` cannot be known at compilation time [E0277]
+        F //~ ERROR cannot find value `F` in this scope [E0425]
+    }
+} //~ ERROR mismatched closing delimiter: `}`
+//~^ ERROR mismatched closing delimiter: `}`
+
+fn main() {}

--- a/src/test/ui/suggestions/constrain-suggest-ice.stderr
+++ b/src/test/ui/suggestions/constrain-suggest-ice.stderr
@@ -1,0 +1,81 @@
+error: mismatched closing delimiter: `}`
+  --> $DIR/constrain-suggest-ice.rs:2:8
+   |
+LL | struct Bug<S>{
+   |              - closing delimiter possibly meant for this
+LL |     A: [(); {
+   |        ^ unclosed delimiter
+...
+LL | }
+   | ^ mismatched closing delimiter
+
+error: mismatched closing delimiter: `}`
+  --> $DIR/constrain-suggest-ice.rs:2:8
+   |
+LL | struct Bug<S>{
+   |              - closing delimiter possibly meant for this
+LL |     A: [(); {
+   |        ^ unclosed delimiter
+...
+LL | }
+   | ^ mismatched closing delimiter
+
+error[E0425]: cannot find value `F` in this scope
+  --> $DIR/constrain-suggest-ice.rs:6:9
+   |
+LL |         F
+   |         ^
+   |
+help: a local variable with a similar name exists
+   |
+LL |         x
+   |         ~
+help: you might be missing a type parameter
+   |
+LL | struct Bug<S, F>{
+   |             +++
+
+error: generic `Self` types are currently not permitted in anonymous constants
+  --> $DIR/constrain-suggest-ice.rs:3:21
+   |
+LL |         let x: [u8; Self::W] = [0; Self::W];
+   |                     ^^^^
+
+error: generic `Self` types are currently not permitted in anonymous constants
+  --> $DIR/constrain-suggest-ice.rs:3:36
+   |
+LL |         let x: [u8; Self::W] = [0; Self::W];
+   |                                    ^^^^
+
+error[E0277]: the size for values of type `S` cannot be known at compilation time
+  --> $DIR/constrain-suggest-ice.rs:3:36
+   |
+LL | struct Bug<S>{
+   |            - this type parameter needs to be `std::marker::Sized`
+LL |     A: [(); {
+LL |         let x: [u8; Self::W] = [0; Self::W];
+   |                                    ^^^^^^^ doesn't have a size known at compile-time
+   |
+note: required by a bound in `Bug`
+  --> $DIR/constrain-suggest-ice.rs:1:12
+   |
+LL | struct Bug<S>{
+   |            ^ required by this bound in `Bug`
+help: consider relaxing the implicit `Sized` restriction
+   |
+LL | struct Bug<S: ?Sized>{
+   |             ++++++++
+
+error[E0392]: parameter `S` is never used
+  --> $DIR/constrain-suggest-ice.rs:1:12
+   |
+LL | struct Bug<S>{
+   |            ^ unused parameter
+   |
+   = help: consider removing `S`, referring to it in a field, or using a marker such as `PhantomData`
+   = help: if you intended `S` to be a const parameter, use `const S: usize` instead
+
+error: aborting due to 7 previous errors
+
+Some errors have detailed explanations: E0277, E0392, E0425.
+For more information about an error, try `rustc --explain E0277`.

